### DISCUSSION TITLE
关于lookup_table_machine_类和DF_Encode类的区别的疑问

### DISCUSSION THREAD

**Discussion by zhangtuo2000** [Open3DV/Xema#25] [Q&A]
请问lookup_table_machine_类和DF_Encode类具体有哪些区别呢？各自有什么特点？

**Comment by vic-w** [comment]
encode相关的代码都是关于解相位和格雷码解码用的。

lookup_table相关的代码是用来进行三维重建的。之所以要用到查找表的方式来做三维重建，原因是我们的单目结构光相机只投竖条纹。
理论上，如果我有相机的(x,y)坐标和投影仪的（x,y)坐标。但这需要投影仪分别投出横条纹和竖条纹才能获得投影仪的(x,y)坐标。
而为了加快成像的速度，我们选择只投影竖条纹，此时我们只能获得投影仪的x坐标。
所以在三维重建的时候，我们只知道相机的(x,y)坐标和投影仪的x坐标，还缺少了投影仪的y坐标。
但根据标定时的对极线约束，投影仪的y坐标和相机的y坐标是相互关联的，所以我们可以从相机那边推算出投影仪的y坐标。
具体的方法，我们使用了两个二维的查找表来实现。
1. 首先将两个镜头转化成光轴平行的针孔相机模型，也就是rectify的过程
2. 用相机的(x,y)坐标查找出相机的归一化坐标(x',y')，这一步就是普通的相机坐标归一化的过程
3. 用相机的y‘和投影仪的x，形成一个二维查找表，查找出投影仪的x'坐标
4. 此时用相机的x’和投影仪的x‘坐标，直接以光轴平行的针孔相机模型重建出三维坐标（X,Y,Z）

我还在另一个帖子里解释过这一过程，[将标定参数生成表的一些疑问。](https://github.com/Open3DV/Xema/discussions/12)
